tbdex/rfq/create.go

```diff
@@ -0,0 +1,220 @@
+		Salt: salt,
+		Payin: &PrivatePaymentDetails{
+			PaymentDetails: payin.PaymentDetails,
+		},
+		Payout: &PrivatePaymentDetails{
+			PaymentDetails: payout.PaymentDetails,
+		},
+		Claims: claims,
+	}
+
+	return h, p, nil
+}
+
+func digestData(salt string, data any) (string, error) {
+	digestible := []any{salt, data}
+
+	byteArray, err := tbdex.DigestJSON(digestible)
+	if err != nil {
+		return "", err
+	}
+
+	encodedString := base64.URLEncoding.EncodeToString(byteArray)
+
+	return encodedString, nil
+}
+
+// PayinMethodWithDetails is used to create the payin method for an RFQ
+type PayinMethodWithDetails struct {
+	Amount         string         `json:"amount"`
+	Kind           string         `json:"kind"`
+	PaymentDetails map[string]any `json:"paymentDetails"`
+}
+
+// PayoutMethodWithDetails is used to create the payout method for an RFQ
+type PayoutMethodWithDetails struct {
+	Kind           string         `json:"kind"`
+	PaymentDetails map[string]any `json:"paymentDetails"`
+}
+
+type rfqHashes struct {
+	PayinHash  string
+	PayoutHash string
+	ClaimsHash string
+}
+
+type createOptions struct {
+	id         string
+	createdAt  time.Time
+	protocol   string
+	externalID string
+	claims     []string
+}
+
+// CreateOption is a function type used to apply options to RFQ creation.
+type CreateOption func(*createOptions)
+
+// ID can be passed to [Create] to provide a custom id.
+func ID(id string) CreateOption {
+	return func(r *createOptions) {
+		r.id = id
+	}
+}
+
+// CreatedAt can be passed to [Create] to provide a custom created at time.
+func CreatedAt(t time.Time) CreateOption {
+	return func(r *createOptions) {
+		r.createdAt = t
+	}
+}
+
+// ExternalID can be passed to [Create] to provide a custom external id.
+func ExternalID(externalID string) CreateOption {
+	return func(r *createOptions) {
+		r.externalID = externalID
+	}
+}
+
+// Claims can be passed to [Create] to provide a custom external id.
+func Claims(claims []string) CreateOption {
+	return func(r *createOptions) {
+		r.claims = claims
+	}
+}
+
+type paymentMethodOptions struct {
+	details map[string]any
+}
+
+// PaymentMethodOption is a function type used to apply options to payment methods.
+type PaymentMethodOption func(*paymentMethodOptions)
+
+// PaymentDetails can be passed to [Payin] to provide arbitrary payment details.
+func PaymentDetails(details map[string]any) PaymentMethodOption {
+	return func(pm *paymentMethodOptions) {
+		pm.details = details
+	}
+}
+
+// Payin can be passed to [Create] to provide a payin method.
+func Payin(amount, kind string, opts ...PaymentMethodOption) PayinMethodWithDetails {
+	s := PayinMethodWithDetails{
+		Amount: amount,
+		Kind:   kind,
+	}
+
+	o := paymentMethodOptions{}
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	s.PaymentDetails = o.details
+
+	return s
+}
+
+// Payout can be passed to [Create] to provide a payout method.
+func Payout(kind string, opts ...PaymentMethodOption) PayoutMethodWithDetails {
```

Comment:
I believe an RFQ will always require a payout to be defined right? perhaps payout should be a parameter in the function signature (as opposed a variadic parameter)? or, at a minimum, should the `Create` function should return an error if a required property is missing

Reply:
oh I see, it already is, so perhaps we can axe these `Payin` and `Payout` functions?

Reply:
different UX for both approaches. using `PayinDetails` as a struct doesn't tell you what fields are required and which ones aren't. `Payin` `Payout` do provide that at the cost of losing context provided via field names